### ISSUE TITLE
`myStream.concurrently(...)` introduces an async boundary for `myStream`

### ISSUE THREAD

**Issue by tbrown1979** [functional-streams-for-scala/fs2#1449]
Example:
```scala
import cats.effect.IO
import fs2._

object Main extends App { 
  val ec = scala.concurrent.ExecutionContext.Implicits.global
  implicit val cs = IO.contextShift(ec)

  Stream.eval(IO(Thread.sleep(10000)))
    .concurrently(Stream.eval(IO(Thread.sleep(10000)))).compile.drain.unsafeRunSync()
}
```
In this example I would expect the first `Thread.sleep(10000)` to evaluate on the `main` thread, and the second `Thread.sleep(10000)` to evaluate on the global EC. However, that's not what happens. Instead, they _both_ evaluate on the global EC and the main thread is just blocking waiting for them to return. 

This was unexpected behavior for me. I assumed that `myStream` would still evaluate in-place instead of being shifted. Maybe this is expected behavior though, so please let me know if this is not actually an issue. 

[Here is a gist showing what each thread is doing](https://gist.github.com/tbrown1979/4b1d58e9116bb343a588ab4dbf0565e9)

**Comment by djspiewak** [comment]
Is this something that actively causes problems? You can always re-shift things within the `Stream.eval`. Normally when errant async boundaries are introduced and it causes problems, it's more of a performance issue, but that doesn't seem to be the concern here.

**Comment by SystemFw** [comment]
Due to the way not only fs2, but also cats-effect works, it's going to be really hard to avoid async boundaries eventually. Matter of fact, we use to have an operation in cats-effect to run an IO up until its first async boundary, and it ended up being so niche that we don't have it anymore.

I think this should be in the "surprising but makes sense" category :)

**Comment by tbrown1979** [comment]
The issue we encountered with this was that we were using `StreamApp` in an old service that uses a jetty server. I wanted the blocking jetty server to run on the `main` thread. Our health server on the other hand was an http4s server. So I did this:

```scala
implicit val ec = scala.concurrent.ExecutionContext.Implicits.global

Stream.eval(IO(jetty.run)).concurrently(http4sHealthServer)
```
and that ended up running the `jetty.run` on the global EC, which was unexpected. For a while we've just had one global thread stuck performing the `jetty.run` which obviously isn't ideal.

Anyway, I don't think this is necessarily an issue, and definitely not a performance one. I was just curious if this was expected behavior, as I did not expect it. Since it sounds like it _is_ expected, I'm fine with closing this issue 😄
